src/server/api/redirect.ts

```diff
@@ -171,18 +180,52 @@ export default async function redirect(req: Express.Request, res: Express.Respon
     if (gaTrackingId) gaLogging(req, res, shortUrl, longUrl)
 
     // Redirect immediately if a crawler is visiting the site
-    if (isCrawler(req.headers["user-agent"] || '')) {
+    if (isCrawler(req.headers['user-agent'] || '')) {
       res.status(302).redirect(longUrl)
       return
-    } else {
-      // Serve transition page
-      // TODO: Get/set a browser cookie so that this isn't served too frequently
+    }
+
+    // Reassure typescript that the session object exists.
+    if (!req.session) {
+      throw new Error('Session object does not exist')
+    }
+
+    if (!req.session.visits || !req.session.visits[shortUrl]) {
+      // This is the first time visiting a/the shortlink.
+      req.session.visits = {
+        ...req.session.visits,
+        [shortUrl]: Math.floor(Date.now()),
+      }
+
+      // Avoid exceeding cookie size limit by pruning
+      // old entries.
+      if (_.size(req.session.visits) > MAX_SHORTURL_COUNT) {
+        // Build inverse dictionary
+        const epochList: number[] = []
+        const lookupTable: EpochToShortUrlMapping = {}
+        Object.keys(req.session.visits).forEach((url) => {
+          const epoch = req.session!.visits[url]
+          lookupTable[epoch] = url
+          epochList.push(epoch)
+        })
+
+        const earliestEpoch = epochList.sort()[0]
+        const earliestShortUrl = lookupTable[earliestEpoch]
+        delete req.session.visits[earliestShortUrl]
+      }
+
       res.status(200)
         .render(TRANSITION_PATH, {
           shortUrl,
           longUrl,
         })
+      return
     }
+
+    // User has visited this shortlink before.
+    // Update LRU cache and redirect.
```

Comment:
how about writing an LRU cache class with a `maxSize` parameter in bytes? This would help to manage complexity in the codebase. Apart from regular get/set functions, you could easily extend to add a `toObject()` function that allows reassignment back to `req.session`.